### PR TITLE
[sc-14313] Duo Catalog Template

### PR DESCRIPTION
**Pull request by arooxa**
Duo Catalog Template - Used SCIM template as reference

[[sc-14313]](https://app.shortcut.com/sgnl/story/20157/productionize-duo-as-an-sor)

### CODE REVIEW

File: duo.sgnl.yaml [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
...dBvniMQkYCbjdJKSFBDISLhblySsMiBbQNsEyS+10MtQfncsi222QNubBvRQmaIb8x6ei+QsWVDhBYivqGQ6pYglNALEv8r4GvbhljmPPTiX20b0TbENwA47wFfAX+3bYllXts2oKWIb8C5VLeA0Iv/lzmUBvw9t7YNsE3oxb+KSBYdQbaa09jcYLQl3II03AHyjjCzYUSL+WTbAIt8BBEJwFXuWEqSeVwuTa5ARALwPnsgO/su4LJIpCTRmJzg58ataC//tW2ARd6DiOSqoH8kadqQtqIb7y5Gue58w3WRnBeclxXo53GxNLnzDddFYvzn6zFL75o1pdUUZSZd5s43XBbJm5KhKJOmDGk7oRdf4FYo+CLrGy6LZLLsy9CLJ8B1I5ZsBy9tG9Agc7/VVZFc6NyxjPGmDdkWQi8+x43S5FPeN1wVSaVcUUqTBVwoTX7In3BRJFVLkRRZClWjS5P8CIUuYfQN10RyC/xY54bQi6dI52KWF7YN2BCFvuGaSF4+cHLVCPi8Zlu2Ep3T/tu2HRug0DdcEsmb0IsfNLkq9OIb3FxZ3UjoxS/oVrVrqW+4IpIralaz8uhql+srh2T5hm4MVyn1DRdEcgt8k5t9+CBCLx4jU30BNWMPeMJ2C+UW+LHMN7ouklvgyToXeQi9eIQIBbgb/LitQkl9o3QqQJdFUvkl1EWEcs+WCqWWb3RVJFfA8SbnaItQ7tkyoVxRM/PsokjesOYqVhFaKM+Q8HAqlL/Q7qhX6hu1Ms/ehoyxwS0q1t1o7h568SRI/EvUgMnDJtNuG7oBfBwk/re0a72ulXyjKyXJBap6ZaX6E3rxZejFfVSIWEoVL36J2kqtzvCfTbGyb2x7SfIJ+KHmWKyNEXrxOEj8CWr08FO71tglbacEif818C3w24ZNWJtvbKtILlDFZyvEkUXPahwFiT9GDY4c4fBGQXpQ5LkWy1Pg9xtO8gJ4HXrx2qYcb5NIrlBTKrdicWstlhPgJEj8Ear6cWzTJptkxPIH1CLcf2V9pcsVShyTTfhGm0VygfrxH4H32yCMIvS8lAmonZNQ22L39cepPQl1NewF8EKvWv8n4DFqce7HFR/TqG/0dGJPNplIDW67vv5s6MUJuSWL9HZ0+Q1Obf1PGuvr0M59Tm6hCV3a/CZvly3f+D+hXH4CuBj5kAAAAABJRU5ErkJggg=="
```
Comment (r-rakshith) [inline duo.sgnl.yaml:3]: Requesting for a shiny new Duo icon
cc: @ivanfong, @jasoncarulli

File: duo.sgnl.yaml [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+      - name: status
+        externalId: status
+        type: String
+      - name: user_id
+        externalId: user_id
+        type: String
+      - name: username
+        externalId: username
+        type: String
+  Group: # https://duo.com/docs/adminapi#groups
```
Comment (r-rakshith) [inline duo.sgnl.yaml:78]:
```suggestion
  # https://duo.com/docs/adminapi#groups
  Group:
```

File: duo.sgnl.yaml [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+      - name: sms_enabled
+        externalId: sms_enabled
+        type: Bool
+      - name: status
+        externalId: status
+        type: String
+      - name: voice_enabled
+        externalId: voice_enabled
+        type: Bool
+  GroupMember: # Fetched from the 'groups' attribute belonging to the Users entity.
```
Comment (r-rakshith) [inline duo.sgnl.yaml:116]: We can move this closer to User entity.

File: duo.sgnl.yaml [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+    attributes:
+      - name: group_id
+        externalId: group_id
+        type: String
+        indexed: true
+        uniqueId: true
+      - name: name
+        externalId: name
+        type: String
+  Phone: # https://duo.com/docs/adminapi#phones
```
Comment (r-rakshith) [inline duo.sgnl.yaml:136]:
```suggestion
  # https://duo.com/docs/adminapi#phones
  Phone:
```

File: duo.sgnl.yaml [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+entities:
+  User: # https://duo.com/docs/adminapi#retrieve-users
+    displayName: User
+    externalId: User
+    description: User entity in Duo
+    syncFrequency: HOURLY
+    syncMinInterval: 1
+    apiCallFrequency: SECONDLY
+    apiCallMinInterval: 1
+    pageSize: 300
```
Comment (r-rakshith) [inline duo.sgnl.yaml:25]: Why is the pagesize 300 everywhere?

File: duo.sgnl.yaml [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+      - name: sms_passcodes_sent
+        externalId: sms_passcodes_sent
+        type: Boolean
+      - name: tampered
+        externalId: tampered
+        type: String
+      - name: type
+        externalId: type
+        type: String
+  PhoneOwner: # Fetched from the 'phones' attribute belonging to the Users entity.
```
Comment (r-rakshith) [inline duo.sgnl.yaml:197]: Move this closer to User entity.

File: duo.sgnl.yaml [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+    pagesOrderedById: false
+    attributes:
+      - name: phone_id
+        externalId: phone_id
+        type: String
+        indexed: true
+        uniqueId: true
+      - name: activated
+        externalId: activated
+        type: Boolean
```
Comment (r-rakshith) [inline duo.sgnl.yaml:150]:
```suggestion
        type: Bool
```

File: duo.sgnl.yaml [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+        type: String
+      - name: predelay
+        externalId: predelay
+        type: String
+      - name: screenlock
+        externalId: screenlock
+        type: String
+      - name: sms_passcodes_sent
+        externalId: sms_passcodes_sent
+        type: Boolean
```
Comment (r-rakshith) [inline duo.sgnl.yaml:190]:
```suggestion
        type: Bool
```

File: duo.sgnl.yaml
Comment (r-rakshith) [inline duo.sgnl.yaml:1]: Let's also add the Endpoint entity to this

**Review by r-rakshith — CHANGES_REQUESTED** [review]
Looks good overall. A few changes required before this is ready to be merged.

File: duo.sgnl.yaml [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+    displayName: Group
+    externalId: Group
+    description: Group entity in Duo
+    syncFrequency: HOURLY
+    syncMinInterval: 1
+    apiCallFrequency: SECONDLY
+    apiCallMinInterval: 1
+    pageSize: 300
+    pagesOrderedById: false
+    attributes:
```
Comment (r-rakshith) [inline duo.sgnl.yaml:88]: This duplicate key makes the YAML invalid.

To make sure that the YAML is valid and contains all the required info, let's also upload this template to 
- a YAML validator
- a SGNL Console (`dev`/`sbx`)

Comment (arooxa) [reply duo.sgnl.yaml:88]: I'm not seeing any duplicates, which key are you referring to?

File: duo.sgnl.yaml (hunk 5 above)
Comment (arooxa) [reply duo.sgnl.yaml:25]: Most endpoints accept a maximum page size of 300, although it's not standardized. There is no limit on the requested page size - eg. a page size of 99999 will be accepted but only return 300 values at most

**Review by richardxgao — CHANGES_REQUESTED** [review]
A few comments on consistency, but overall looks good!

File: duo.sgnl.yaml (hunk 5 above)
Comment (r-rakshith) [reply duo.sgnl.yaml:26]: Let's stick to page size 100 as a standard. 
We can document the max allowed page size and possibly add validation on the adapter end if we know the limits on page size.

File: duo.sgnl.yaml (hunk 9 above)
Comment (r-rakshith) [reply duo.sgnl.yaml:132]: Sure, we can sync on this offline.

File: duo.sgnl.yaml [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+        type: String
+      - name: platform
+        externalId: platform
+        type: String
+      - name: postdelay
+        externalId: postdelay
+        type: String
+      - name: predelay
+        externalId: predelay
+        type: String
```
[the remark this answers is not in the document]
Comment (r-rakshith) [reply duo.sgnl.yaml:208]: I had the same question during the first pass of review. These are strings. From the response samples, it'd be stored as `"0"`.

File: duo.sgnl.yaml [hunk 11]
```diff
@@ -0,0 +1,338 @@
...
+        externalId: group_id
+        type: String
+        indexed: true
+        uniqueId: true
+      - name: name
+        externalId: name
+        type: String
+  # Fetched from the 'phones' attribute belonging to the Users entity.
+  PhoneOwner:
+    parent: users
```
[the remark this answers is not in the document]
Comment (arooxa) [reply duo.sgnl.yaml:102]: Yup, parent field should be the external id of the parent identity. I changed the User externalID earlier and forgot to update this, good catch!

File: duo.sgnl.yaml [hunk 12]
```diff
@@ -0,0 +1,338 @@
...
+        externalId: username
+        type: String
+relationships:
+  ParentOfGroupMember:
+    childEntity: groups
+  ParentOfPhoneOwner:
+    childEntity: phones
+  MemberOfGroup:
+    name: MemberOf
+    fromAttribute: groups.group_id
```
[the remark this answers is not in the document]
Comment (arooxa) [reply duo.sgnl.yaml:322]: @r-rakshith the SCIM catalog also uses the external id to reference attributes in the entity relationships, is this expected?

[the remark this answers is not in the document]
Comment (richardxgao) [reply duo.sgnl.yaml:322]: If SCIM also uses this format, and it works, then I'm OK with leaving as-is. This is not a major issue.

File: duo.sgnl.yaml (hunk 9 above)
Comment (arooxa) [reply duo.sgnl.yaml:135]: Verified that this works with the console. Next step is to wait for the duo adapter to be merged and test before merging this catalog right?

**Review by richardxgao — APPROVED** [review]
LGTM, but one minor comment is not addressed: https://github.com/SGNL-ai/catalog-sgnl-templates/pull/42#discussion_r1478368876. 

If you need help testing the Duo adapter let me or Rakshith know.

**Review by r-rakshith — DISMISSED** [review]
LGTM. 
Let's test this catalog on Duo adapter once it is released to `DEV` and then merge this.

File: duo.sgnl.yaml [hunk 13]
```diff
@@ -0,0 +1,341 @@
...
+        type: String
+      - name: username
+        externalId: username
+        type: String
+  # Fetched from the 'groups' attribute belonging to the Users entity.
+  GroupMember:
+    parent: User
+    displayName: GroupMember
+    externalId: groups
+    description: Entity representing DUO's user and group membership
```
Comment (r-rakshith) [inline duo.sgnl.yaml:87]:
```suggestion
    description: Entity representing the Duo groups each user belongs to

File: duo.sgnl.yaml [hunk 14]
```diff
@@ -0,0 +1,341 @@
...
+        uniqueId: true
+      - name: name
+        externalId: name
+        type: String
+  # Fetched from the 'phones' attribute belonging to the Users entity.
+  PhoneOwner:
+    parent: User
+    displayName: PhoneOwner
+    externalId: phones
+    description: Entity representing DUO's user and phone ownership
```
Comment (r-rakshith) [inline duo.sgnl.yaml:108]:
```suggestion
    description: Entity representing the Phone owned by the User
```

File: duo.sgnl.yaml [hunk 15]
```diff
@@ -0,0 +1,341 @@
...
+        externalId: status
+        type: String
+      - name: voice_enabled
+        externalId: voice_enabled
+        type: Bool
+  # https://duo.com/docs/adminapi#phones
+  Phone:
+    displayName: Phone
+    externalId: Phone
+    description: Duo Phone Entity
```
Comment (r-rakshith) [inline duo.sgnl.yaml:166]:
```suggestion
    description: Phone entity in Duo
```

File: duo.sgnl.yaml [hunk 16]
```diff
@@ -0,0 +1,341 @@
...
+        externalId: tampered
+        type: String
+      - name: type
+        externalId: type
+        type: String
+  # https://duo.com/docs/adminapi#retrieve-endpoints
+  Endpoint:
+    displayName: Endpoint
+    externalId: Endpoint
+    description: Duo Endpoint Entity
```
Comment (r-rakshith) [inline duo.sgnl.yaml:228]:
```suggestion
    description Endpoint entity in Duo
```

**Review by r-rakshith — COMMENTED** [review]
Also, some changes to description that addresses Richards' comments.

**Review by richardxgao — APPROVED** [review]
The remaining open comment can be addressed in a separate PR.

File: duo.sgnl.yaml (hunk 9 above)
Comment (arooxa) [reply duo.sgnl.yaml:135]: finished testing

**Review by r-rakshith — APPROVED** [review]
LGTM. Thanks!

File: duo.sgnl.yaml [hunk 17]
```diff
@@ -0,0 +1,341 @@
...
...3BhY2l0eToxO2ZpbGwtcnVsZTpub256ZXJvO3N0cm9rZTpub25lIgogICAgICAgICBpZD0icGF0aDQ1IgogICAgICAgICBpbmtzY2FwZTpjb25uZWN0b3ItY3VydmF0dXJlPSIwIiAvPjxwYXRoCiAgICAgICAgIGQ9Ik0gODM4My42OCwtMC4wMTE3MTg4IEMgNzUyMC4zLC0wLjAxMTcxODggNjgxNS40NSw2NzguNjYgNjc3My40NSwxNTMxLjYzIEggOTk5My45MyBDIDk5NTEuOTIsNjc4LjY2IDkyNDcuMDcsLTAuMDExNzE4OCA4MzgzLjY4LC0wLjAxMTcxODgiCiAgICAgICAgIHN0eWxlPSJmaWxsOiM3YmNkNTQ7ZmlsbC1vcGFjaXR5OjE7ZmlsbC1ydWxlOm5vbnplcm87c3Ryb2tlOm5vbmUiCiAgICAgICAgIGlkPSJwYXRoNDciCiAgICAgICAgIGlua3NjYXBlOmNvbm5lY3Rvci1jdXJ2YXR1cmU9IjAiIC8+PHBhdGgKICAgICAgICAgZD0iTSAzMzg1LjcyLDMyMjQuNDkgViAxNjEyLjI0IEMgMzM4NS43Miw3NDguODU5IDQwNjQuMzksNDQuMDExNyA0OTE3LjM2LDIgViAzMjI0LjQ5IEggMzM4NS43MiIKICAgICAgICAgc3R5bGU9ImZpbGw6IzYzYzQzZjtmaWxsLW9wYWNpdHk6MTtmaWxsLXJ1bGU6bm9uemVybztzdHJva2U6bm9uZSIKICAgICAgICAgaWQ9InBhdGg0OSIKICAgICAgICAgaW5rc2NhcGU6Y29ubmVjdG9yLWN1cnZhdHVyZT0iMCIgLz48cGF0aAogICAgICAgICBkPSJNIDY2MTAuMjIsLTAuMDExNzE4OCBIIDUwNzguNTggViAzMjI0LjQ5IEggNjYxMC4yMiBWIDE2MTIuMjQgLTAuMDExNzE4OCIKICAgICAgICAgc3R5bGU9ImZpbGw6IzdiY2Q1NDtmaWxsLW9wYWNpdHk6MTtmaWxsLXJ1bGU6bm9uemVybztzdHJva2U6bm9uZSIKICAgICAgICAgaWQ9InBhdGg1MSIKICAgICAgICAgaW5rc2NhcGU6Y29ubmVjdG9yLWN1cnZhdHVyZT0iMCIgLz48L2c+PC9nPjwvc3ZnPg=="
```
Comment (r-rakshith) [inline duo.sgnl.yaml:3]:
```suggestion
icon: "PHN2ZyB3aWR0aD0iNjQiIGhlaWdodD0iNjQiIHZpZXdCb3g9IjAgMCA2NCA2NCIgZmlsbD0ibm9uZSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj4KPHJlY3Qgd2lkdGg9IjY0IiBoZWlnaHQ9IjY0IiBmaWxsPSIjNkRCRTUwIi8+CjxnIG9wYWNpdHk9IjAuODUiPgo8cGF0aCBkPSJNMjMuNDcyIDMxLjc2OEg4VjM5LjcwNEgxNS43NDRDMTkuODg4IDM5LjcwNCAyMy4yNzIgMzYuMTkyIDIzLjQ3MiAzMS43NjhaIiBmaWxsPSJ3aGl0ZSIvPgo8L2c+CjxwYXRoIGQ9Ik0xNS43NDQgMjNIOFYzMC45MzZIMjMuNDhDMjMuMjcyIDI2LjUxMiAxOS44ODggMjMgMTUuNzQ0IDIzWiIgZmlsbD0id2hpdGUiLz4KPHBhdGggZD0iTTQ4LjI2NCAyM0M0NC4xMiAyMyA0MC43MjggMjYuNTIgNDAuNTI4IDMwLjkzNkg1NS45OTJDNTUuOCAyNi41MTIgNTIuNDA4IDIzIDQ4LjI2NCAyM1oiIGZpbGw9IndoaXRlIi8+CjxnIG9wYWNpdHk9IjAuODUiPgo8cGF0aCBkPSJNNTYgMzEuNzY4SDQwLjUyOEM0MC43MjggMzYuMTkyIDQ0LjExMiAzOS43MDQgNDguMjY0IDM5LjcwNEM1Mi40MDggMzkuNzA0IDU1LjggMzYuMTkyIDU2IDMxLjc2OFoiIGZpbGw9IndoaXRlIi8+CjwvZz4KPHBhdGggZD0iTTI0LjI1NiAyM1YzMS4zNTJDMjQuMjU2IDM1LjgyNCAyNy41MTIgMzkuNDggMzEuNjE2IDM5LjY5NlYyM0gyNC4yNTZaIiBmaWxsPSJ3aGl0ZSIvPgo8ZyBvcGFjaXR5PSIwLjg1Ij4KPHBhdGggZD0iTTM5Ljc0NCAyM0gzMi4zOTJWMzkuNzA0SDM5Ljc0NFYzMS4zNTJWMjNaIiBmaWxsPSJ3aGl0ZSIvPgo8L2c+Cjwvc3ZnPg=="
```

**Review by r-rakshith — APPROVED** [review]

**Review by richardxgao — APPROVED** [review]

**Pull request merged**